README.md
# cha-price

[![npm version](https://img.shields.io/npm/v/cha-price.svg?style=flat-square)](https://www.npmjs.com/package/cha-price)
[![npm downloads](https://img.shields.io/npm/dm/cha-price.svg?style=flat-square)](https://www.npmjs.com/package/cha-price)
[![Build Status](https://img.shields.io/travis/lgaticaq/cha-price.svg?style=flat-square)](https://travis-ci.org/lgaticaq/cha-price)
[![Coverage Status](https://img.shields.io/coveralls/lgaticaq/cha-price/master.svg?style=flat-square)](https://coveralls.io/github/lgaticaq/cha-price?branch=master)
[![Code Climate](https://img.shields.io/codeclimate/github/lgaticaq/cha-price.svg?style=flat-square)](https://codeclimate.com/github/lgaticaq/cha-price)
[![devDependency Status](https://img.shields.io/david/dev/lgaticaq/cha-price.svg?style=flat-square)](https://david-dm.org/lgaticaq/cha-price#info=devDependencies)

> Get price of chaucha

## Installation

```bash
npm i -S cha-price
```

## Use

[Try on RunKit](https://npm.runkit.com/cha-price)
```js
const cha = require('cha-price');

// Get from https://orionx.io/developers
const apiKey = 'your-api-key'
const secretKey = 'your-secret-key'

cha(apiKey, secretKey)
  .then(console.log)
  .catch(console.error);
```

Result:
```js
XXXXX // a number
```

## Licencia

[MIT](https://tldrlegal.com/license/mit-license)


CHANGELOG.md
## [2.0.1](https://github.com/lgaticaq/cha-price/compare/v2.0.0...v2.0.1) (2018-08-23)


### Bug Fixes

* **orionx:** fix HMAC ([3bfc9d9](https://github.com/lgaticaq/cha-price/commit/3bfc9d9))

## 2.0.0 (2018-05-01)

##### Breaking Changes

* **orionx:**  migrate to new api ([363ab2f4](https://github.com/lgaticaq/cha-price/commit/363ab2f463235216af4433db80f7435dc5e7fd7d))

##### Chores

* **package:**
  *  update dependencies ([940e2a1f](https://github.com/lgaticaq/cha-price/commit/940e2a1f61bdd48a8fed851d19369d797c648a69))
  *  add dist files ([ec41225d](https://github.com/lgaticaq/cha-price/commit/ec41225d0cf0a97e27756cdeee372b1c821286b4))
* **git:**  add .vscode to ignored ([bdc863d7](https://github.com/lgaticaq/cha-price/commit/bdc863d72ff012f9ead3dd579d3ecffb7e6984a6))
* **nvm:**  update node to v10 ([a2fc627b](https://github.com/lgaticaq/cha-price/commit/a2fc627b88d4643d61b49638f82b1081e9d16f7b))
* **deps:**
  *  update dependency lint-staged to v7.0.5 ([64ddbc8d](https://github.com/lgaticaq/cha-price/commit/64ddbc8d9b5a0975110d955dbcd347413928f9ca))
  *  update dependency eslint-plugin-standard to v3.1.0 ([3e8f8a78](https://github.com/lgaticaq/cha-price/commit/3e8f8a78a06ef9e5890f40ecdd5bbb80489c726b))
  *  update dependency mocha to v5.1.1 ([6cfde2b8](https://github.com/lgaticaq/cha-price/commit/6cfde2b88cf5a9e17a1b16c7bc473c6d3510f971))
  *  update dependency mocha to v5.1.0 ([0b4a5721](https://github.com/lgaticaq/cha-price/commit/0b4a5721b8647b2a163016a01d856beb52ee2697))
  *  update dependency eslint-plugin-import to v2.11.0 ([e6a56202](https://github.com/lgaticaq/cha-price/commit/e6a5620288053cd4e7d468e5be53722fb6aa590a))
  *  update dependency nock to v9.2.5 ([aea06b0d](https://github.com/lgaticaq/cha-price/commit/aea06b0d7793591ed5977a41d9f3d91e16391048))
  *  update dependency nock to v9.2.4 ([36e1cb43](https://github.com/lgaticaq/cha-price/commit/36e1cb4321954d53574c0e605cf3e537362ae392))
  *  update dependency prettier-standard to v8.0.1 ([817cf57d](https://github.com/lgaticaq/cha-price/commit/817cf57da7811863556d411130aa7891850c9594))
  *  update dependency lint-staged to v7.0.4 ([3a46731c](https://github.com/lgaticaq/cha-price/commit/3a46731c15fd528aa32c10e458dab9e838db228f))
  *  update dependency lint-staged to v7.0.3 ([30af2ff9](https://github.com/lgaticaq/cha-price/commit/30af2ff90af3d3a01178220beec7d740c10ed956))
  *  update dependency lint-staged to v7.0.2 ([15086d74](https://github.com/lgaticaq/cha-price/commit/15086d7416ceb35d18c169b5fdda7a9d312c2b61))
  *  update dependency lint-staged to v7.0.1 ([c979db8c](https://github.com/lgaticaq/cha-price/commit/c979db8c4da6613436aaf3572a1cdf6b94aee827))
  *  update dependency eslint-plugin-import to v2.10.0 ([5d2cd243](https://github.com/lgaticaq/cha-price/commit/5d2cd2434f4eee6c82e7b61247f721c36c3ac19b))
  *  update dependency generate-changelog to v1.7.1 ([3724516e](https://github.com/lgaticaq/cha-price/commit/3724516ed38da0962e5fccdb4d4941e205bd2880))
  *  update dependency mocha to v5.0.5 ([da7aeb51](https://github.com/lgaticaq/cha-price/commit/da7aeb518f3fcc4c30159dcb5bcacc236f915f55))
  *  update dependency eslint to v4.19.1 ([28a12ca3](https://github.com/lgaticaq/cha-price/commit/28a12ca3ff3480bec6b22b4bb4deba88fac025f9))
  *  update dependency eslint to v4.19.0 ([be96fad1](https://github.com/lgaticaq/cha-price/commit/be96fad1798ef54648e603616aa11f2b5ffc8c99))
  *  update dependency eslint-plugin-promise to v3.7.0 ([e29f46c9](https://github.com/lgaticaq/cha-price/commit/e29f46c9d627c2ff3063950cf1171983b043407d))
  *  update dependency mocha to v5.0.4 ([0c651761](https://github.com/lgaticaq/cha-price/commit/0c651761000f96f9a38eb0eb7634114bef75daa7))
  *  update dependency mocha to v5.0.3 ([a66fde38](https://github.com/lgaticaq/cha-price/commit/a66fde38494fa15a282873a708c429f707d84104))
  *  update dependency mocha to v5.0.2 ([70779815](https://github.com/lgaticaq/cha-price/commit/70779815cd9b4f4e2f8feb75162523bdec81e2c1))
  *  update dependency eslint to v4.18.2 ([300002bc](https://github.com/lgaticaq/cha-price/commit/300002bc38d96a34cace4e7a303fb0419509720c))
  *  update dependency nock to v9.2.3 ([4440bc8d](https://github.com/lgaticaq/cha-price/commit/4440bc8d7ce8c57cb0fa2f5b3d7545951193cce3))
  *  update dependency nock to v9.2.2 ([d818924d](https://github.com/lgaticaq/cha-price/commit/d818924dd4c0aabc013029a06252242cb4526852))
  *  update dependency nock to v9.2.1 ([cb60e37a](https://github.com/lgaticaq/cha-price/commit/cb60e37a5abecacbde4e00a37b34e436560c107e))
  *  update dependency nock to v9.2.0 ([18a16790](https://github.com/lgaticaq/cha-price/commit/18a1679070f36b8b91cd0b9a918e787c7e5720a6))
  *  update dependency lint-staged to v7.0.0 ([dc5c515a](https://github.com/lgaticaq/cha-price/commit/dc5c515ab880f0dd6ea2cbdad41417755c569524))
  *  update dependency nock to v9.1.10 ([ce5841d2](https://github.com/lgaticaq/cha-price/commit/ce5841d2fd2522e4813135770b84988fd5b35b1f))
  *  update dependency eslint-plugin-node to v6.0.1 ([c06f9523](https://github.com/lgaticaq/cha-price/commit/c06f9523d79fde740a15dcc6413d5cda186eaba1))
  *  update dependency eslint-plugin-import to v2.9.0 ([4335be53](https://github.com/lgaticaq/cha-price/commit/4335be53e908c316f62d4158459c6ff3e3864a64))
  *  update dependency nock to v9.1.9 ([3e8bb629](https://github.com/lgaticaq/cha-price/commit/3e8bb629b5a09192c51d0c1d4099691eaacff9c2))
  *  update dependency eslint to v4.18.1 ([6c27a379](https://github.com/lgaticaq/cha-price/commit/6c27a3798f1bba6ecd8d32f4bbbbac153251ecb3))
  *  update dependency eslint-config-standard to v11.0.0 ([24556623](https://github.com/lgaticaq/cha-price/commit/245566232b6d5b142847f9508614291374156009))
  *  update dependency mocha to v5.0.1 ([b0eec6f4](https://github.com/lgaticaq/cha-price/commit/b0eec6f4f5354b4964898b293f09987c5206c4ac))
  *  update dependency eslint-plugin-node to v6.0.0 ([19dd4448](https://github.com/lgaticaq/cha-price/commit/19dd4448b39925ef1e82f956d5d90ff76c23a787))

test/test.js
'use strict'

const { describe, it, beforeEach } = require('mocha')
const { expect } = require('chai')
const nock = require('nock')

const lib = require('../src')
const apiKey = '5WGDN4rD3Eqpz9tnMQuqyRCfqZUP4ccJsb'
const secretKey = 'B7EWrck6QaJ8mPa5jYCqKNaagKiHDtq9LT'

describe('cha', () => {
  describe('valid', () => {
    beforeEach(() => {
      nock('https://api2.orionx.io', {
        reqheaders: {
          'Content-Type': 'application/json',
          'X-ORIONX-TIMESTAMP': timestamp => /^\d{10}\.\d+$/.test(timestamp),
          'X-ORIONX-APIKEY': '5WGDN4rD3Eqpz9tnMQuqyRCfqZUP4ccJsb',
          'X-ORIONX-SIGNATURE': signature => /^\w{128}$/.test(signature)
        }
      })
        .post('/graphql', {
          query: '{ market(code: "CHACLP") { lastTrade { price }}}'
        })
        .reply(200, {
          data: {
            market: {
              lastTrade: {
                price: 10000
              }
            }
          }
        })
    })

    it('should return a price', done => {
      lib(apiKey, secretKey).then(price => {
        expect(price).to.eql(10000)
        done()
      })
    })
  })

  describe('valid json with empty price', () => {
    beforeEach(() => {
      nock('https://api2.orionx.io', {
        reqheaders: {
          'Content-Type': 'application/json',
          'X-ORIONX-TIMESTAMP': timestamp => /^\d{10}\.\d+$/.test(timestamp),
          'X-ORIONX-APIKEY': '5WGDN4rD3Eqpz9tnMQuqyRCfqZUP4ccJsb',
          'X-ORIONX-SIGNATURE': signature => /^\w{128}$/.test(signature)
        }
      })
        .post('/graphql', {
          query: '{ market(code: "CHACLP") { lastTrade { price }}}'
        })
        .reply(200, {})
    })

    it('should return a null', done => {
      lib(apiKey, secretKey).then(price => {
        expect(price).to.be.an('null')
        done()
      })
    })
  })

  describe('invalid json', () => {
    beforeEach(() => {
      nock('https://api2.orionx.io', {
        reqheaders: {
          'Content-Type': 'application/json',
          'X-ORIONX-TIMESTAMP': timestamp => /^\d{10}\.\d+$/.test(timestamp),
          'X-ORIONX-APIKEY': '5WGDN4rD3Eqpz9tnMQuqyRCfqZUP4ccJsb',
          'X-ORIONX-SIGNATURE': signature => /^\w{128}$/.test(signature)
        }
      })
        .post('/graphql', {
          query: '{ market(code: "CHACLP") { lastTrade { price }}}'
        })
        .reply(200, '{10000}')
    })

    it('should return an error', done => {
      lib(apiKey, secretKey).catch(err => {
        expect(err).to.be.an('error')
        done()
      })
    })
  })

  describe('wrong status code', () => {
    beforeEach(() => {
      nock('https://api2.orionx.io', {
        reqheaders: {
          'Content-Type': 'application/json',
          'X-ORIONX-TIMESTAMP': timestamp => /^\d{10}\.\d+$/.test(timestamp),
          'X-ORIONX-APIKEY': '5WGDN4rD3Eqpz9tnMQuqyRCfqZUP4ccJsb',
          'X-ORIONX-SIGNATURE': signature => /^\w{128}$/.test(signature)
        }
      })
        .post('/graphql', {
          query: '{ market(code: "CHACLP") { lastTrade { price }}}'
        })
        .reply(301)
    })

    it('should return an error', done => {
      lib(apiKey, secretKey).catch(err => {
        expect(err).to.be.an('error')
        done()
      })
    })
  })

  describe('server error', () => {
    beforeEach(() => {
      nock('https://api2.orionx.io', {
        reqheaders: {
          'Content-Type': 'application/json',
          'X-ORIONX-TIMESTAMP': timestamp => /^\d{10}\.\d+$/.test(timestamp),
          'X-ORIONX-APIKEY': '5WGDN4rD3Eqpz9tnMQuqyRCfqZUP4ccJsb',
          'X-ORIONX-SIGNATURE': signature => /^\w{128}$/.test(signature)
        }
      })
        .post('/graphql', {
          query: '{ market(code: "CHACLP") { lastTrade { price }}}'
        })
        .replyWithError('Server error')
    })

    it('should return an error', done => {
      lib(apiKey, secretKey).catch(err => {
        expect(err).to.be.an('error')
        done()
      })
    })
  })
})


src/browser.js
'use strict'

/* global fetch */

import jsSHA from 'jssha'

export default (apiKey, secretKey) => {
  const timestamp = new Date().getTime() / 1000
  const postData = JSON.stringify({
    query: '{ market(code: "CHACLP") { lastTrade { price }}}'
  })
  const sha = new jsSHA('SHA-512', 'TEXT') // eslint-disable-line
  sha.setHMACKey(secretKey, 'TEXT')
  sha.update(`${timestamp}${postData}`)
  const signature = sha.getHMAC('HEX')
  return fetch('https://api2.orionx.io/graphql', {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
      'Content-Length': postData.length,
      'X-ORIONX-TIMESTAMP': timestamp,
      'X-ORIONX-APIKEY': apiKey,
      'X-ORIONX-SIGNATURE': signature
    },
    body: postData
  })
    .then(res => res.json())
    .then(res => res.data.market.lastTrade.price)
}


src/index.js
'use strict'

const https = require('https')
const crypto = require('crypto')

module.exports = (apiKey, secretKey) => {
  return new Promise((resolve, reject) => {
    if (!apiKey || !secretKey) {
      return reject(new Error('Empty apiKey or secretKey'))
    }
    const timestamp = new Date().getTime() / 1000
    const postData = JSON.stringify({
      query: '{ market(code: "CHACLP") { lastTrade { price }}}'
    })
    const options = {
      hostname: 'api2.orionx.io',
      port: 443,
      path: '/graphql',
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        'Content-Length': Buffer.byteLength(postData),
        'X-ORIONX-TIMESTAMP': timestamp,
        'X-ORIONX-APIKEY': apiKey,
        'X-ORIONX-SIGNATURE': crypto
          .createHmac('sha512', secretKey)
          .update(`${timestamp}${postData}`)
          .digest('hex')
      }
    }
    const req = https.request(options, res => {
      if (res.statusCode !== 200) {
        reject(new Error(`Request Failed. Status Code: ${res.statusCode}`))
      } else {
        res.setEncoding('utf8')
        let rawData = ''
        res.on('data', chunk => {
          rawData += chunk
        })
        res.on('end', () => {
          try {
            const json = JSON.parse(rawData)
            if (json.data && json.data.market && json.data.market.lastTrade) {
              resolve(json.data.market.lastTrade.price)
            } else {
              resolve(null)
            }
          } catch (err) {
            reject(err)
          }
        })
      }
    })
    req.on('error', err => reject(err))
    req.write(postData)
    req.end()
  })
}
